File: packages/java/examples/04_multi_agent/RunDemo.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

/**
 * Multi-Agent Demo - Orchestrates multiple agents.
 *
 * First start the agents:
 *   ./gradlew run -PmainClass=multi_agent.FinanceAgent
 *   ./gradlew run -PmainClass=multi_agent.ReporterAgent
 * Then run: ./gradlew run -PmainClass=multi_agent.RunDemo
 */
class RunDemo {
    private static final HttpClient http = HttpClient.newHttpClient();
    private static final ObjectMapper mapper = new ObjectMapper();

    public static void main(String[] args) throws Exception {
        System.out.println("🚀 Multi-Agent Demo\n");

        try {
            // Step 1: Get stock prices
            System.out.println("📊 Getting stock prices...");
            for (String symbol : new String[]{"AAPL", "GOOGL", "MSFT"}) {
                var result = callAgent("http://localhost:8788", "get_stock_price", Map.of("symbol", symbol));
                System.out.printf("   %s: $%s%n", symbol, result.get("price"));
            }

            // Step 2: Calculate portfolio
            System.out.println("\n💰 Calculating portfolio...");
            var portfolio = callAgent("http://localhost:8788", "get_portfolio_value",
                Map.of("holdings", Map.of("AAPL", 10, "GOOGL", 5, "MSFT", 8)));
            System.out.printf("   Total: $%.2f%n", ((Number) portfolio.get("total_value")).doubleValue());

            // Step 3: Generate report
            System.out.println("\n📝 Generating report...");
            var report = callAgent("http://localhost:8789", "generate_summary",
                Map.of("data", Map.of("portfolio_value", portfolio.get("total_value"), "stocks", 3), "format", "markdown"));
            System.out.println("\n" + report.get("report"));

            System.out.println("✅ Demo complete!");
        } catch (Exception e) {
            System.err.println("❌ Error: " + e.getMessage());
            System.out.println("\nStart agents first:\n  ./gradlew run -PmainClass=multi_agent.FinanceAgent\n  ./gradlew run -PmainClass=multi_agent.ReporterAgent");
        }
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> callAgent(String url, String skill, Map<String, Object> params) throws Exception {
        var body = mapper.writeValueAsString(Map.of("jsonrpc", "2.0", "id", "1", "method", "message/send",
            "params", Map.of("message", Map.of("role", "user", "parts",
                new Object[]{Map.of("kind", "text", "text", mapper.writeValueAsString(Map.of("skill", skill, "params", params)))}))));
        var request = HttpRequest.newBuilder().uri(URI.create(url))
            .header("Content-Type", "application/json").POST(HttpRequest.BodyPublishers.ofString(body)).build();
        var response = http.send(request, HttpResponse.BodyHandlers.ofString());
        JsonNode json = mapper.readTree(response.body());
        String text = json.at("/result/parts/0/text").asText();
        return mapper.readValue(text, Map.class);
    }
}

File: packages/typescript/examples/04_multi_agent/run_demo.ts
/**
 * Multi-Agent Demo - Orchestrates multiple agents.
 *
 * First start: npx ts-node examples/04_multi_agent/finance_agent.ts
 *              npx ts-node examples/04_multi_agent/reporter_agent.ts
 * Then run:    npx ts-node examples/04_multi_agent/run_demo.ts
 */
async function callAgent(agentUrl: string, skill: string, params: Record<string, unknown>) {
  const response = await fetch(agentUrl, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      jsonrpc: '2.0', id: '1', method: 'message/send',
      params: { message: { role: 'user', parts: [{ kind: 'text', text: JSON.stringify({ skill, params }) }] } },
    }),
  });
  const result = await response.json() as any;
  const textPart = result.result?.parts?.find((p: any) => p.kind === 'text' || p.type === 'text');
  return textPart ? JSON.parse(textPart.text) : result.result;
}

async function runDemo() {
  console.log('🚀 Multi-Agent Demo\n');
  try {
    console.log('📊 Getting stock prices...');
    for (const symbol of ['AAPL', 'GOOGL', 'MSFT']) {
      const result = await callAgent('http://localhost:8788', 'get_stock_price', { symbol });
      console.log('   ' + symbol + ': $' + result.price);
    }

    console.log('\n💰 Calculating portfolio...');
    const portfolio = await callAgent('http://localhost:8788', 'get_portfolio_value', { holdings: { AAPL: 10, GOOGL: 5, MSFT: 8 } });
    console.log('   Total: $' + portfolio.total_value.toFixed(2));

    console.log('\n📝 Generating report...');
    const report = await callAgent('http://localhost:8789', 'generate_summary', { data: { portfolio_value: portfolio.total_value, stocks: 3 }, format: 'markdown' });
    console.log('\n' + report.report);

    console.log('\n✅ Demo complete!');
  } catch (e: any) {
    console.error('❌ Error: ' + e.message);
    console.log('\nStart agents first:\n  npx ts-node examples/04_multi_agent/finance_agent.ts\n  npx ts-node examples/04_multi_agent/reporter_agent.ts');
  }
}

runDemo();
